frontend/lib/src/components/elements/Dialog/Dialog.tsx

```diff
@@ -0,0 +1,142 @@
+import IsSidebarContext from "@streamlit/lib/src/components/core/IsSidebarContext"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+
+import { StyledDialogContent } from "./styled-components"
+
+export interface Props {
+  element: BlockProto.Dialog
+  widgetMgr: WidgetStateManager
+}
+
+type DialogWidth = "small" | "large"
+
+function parseWidthConfig(width: DialogWidth, theme: EmotionTheme): string {
+  if (width === "large") {
+    // this is the same width including padding as the AppView container is using for all inner elements
+    // the padding is added to the ModalBody and subtracted from the total width here
+    // As of writing this: total width=752px (47rem), padding left and right each=24px (1.5rem) => content width = 704px
+    return `calc(${theme.sizes.contentMaxWidth} + 2*${theme.spacing.sm})`
+  }
+
+  return SIZE.default
+}
+
+function isDialogWidth(str: string | null | undefined): str is DialogWidth {
+  return str === "small" || str === "large"
+}
+
+const Dialog: React.FC<React.PropsWithChildren<Props>> = ({
+  element,
+  children,
+  widgetMgr,
+}): ReactElement => {
+  const { title, dismissible, width, isOpen: initialIsOpen } = element
+  const [isOpen, setIsOpen] = useState<boolean>(false)
+
+  useEffect(() => {
+    // release the lock properly when isOpen changed / the component unmounts
+    return () => widgetMgr.releaseDialogLock()
+  }, [widgetMgr])
+
+  useEffect(() => {
+    // Only apply the open state if it was actually set in the proto.
+    if (notNullOrUndefined(initialIsOpen)) {
+      // only open when the lock can be acquired
+      const lockAcquired = widgetMgr.tryAcquireDialogLock()
+      setIsOpen(initialIsOpen && lockAcquired)
+    }
+  }, [initialIsOpen, widgetMgr])
+
+  const { activeTheme } = React.useContext(LibContext)
+  const isInSidebar = React.useContext(IsSidebarContext)
+
+  const ThemedModal = useMemo(() => {
+    return function ThemedModal({ children }: { children: ReactNode }) {
+      if (isInSidebar) {
+        return (
+          <ThemeProvider
+            theme={activeTheme.emotion}
+            baseuiTheme={activeTheme.basewebTheme}
+          >
+            {children}
+          </ThemeProvider>
+        )
+      }
+
+      return <>{children}</>
+    }
+  }, [activeTheme.emotion, activeTheme.basewebTheme, isInSidebar])
+
+  const size: string = useMemo(
+    () =>
+      parseWidthConfig(
+        isDialogWidth(width) ? width : "small",
+        activeTheme.emotion
+      ),
+    [width, activeTheme]
+  )
+
+  return (
+    <ThemedModal>
+      <Modal
+        isOpen={isOpen}
+        closeable={dismissible}
+        onClose={() => {
+          setIsOpen(false)
+          widgetMgr.releaseDialogLock()
+        }}
+        size={size}
+        overrides={{
+          Dialog: {
+            style: {
+              // make sure the modal is not too small on mobile
+              minWidth: "20rem",
```

Comment:
Seems like a theme thing we should apply to all of our dialogs?

Reply:
absolutely agree!